### ISSUE TITLE
PDB broken

### ISSUE THREAD

**Issue by aaxu** [google/ci_edit#92]
It seems that at some point, one of the commits broke pdb. I've been sometimes using `import pdb; pdb.set_trace()` to debug the program, but now it's really difficult for the debugger to take in user input. What I mean is that you can maybe press a key 5 times and only one of them will be received by the debugger. Not sure why this is happening only recently.

**Comment by dschuyler** [comment]
My first guess is that it would be related to my moving much of the code to a background thread. There is also code that will look for groups of keypresses to process (up to five). That code has also been changing. I'll try using pdb.set_trace() to see if I can see/fix something. Is there a particular place in ci_edit where the set_trace() is being called from?

**Comment by aaxu** [comment]
I was using it in redo() and undo(). How do you usually debug this program?
I find it somewhat difficult to debug curses applications. Any tips would
be appreciated :D

**Comment by dschuyler** [comment]
I've added a --singleThread command line option to change the editor back to using a single thread (no background thread). This allows pdb to work properly.

I debug using app.log.info(), app.log.debug(), app.log.exception() and so on. Essentially what's called 'printf debugging'. When I use a debugger, I tend to use gdb, pdb, or lldb.

For debugging I generally follow a 'whatever works' approach. There's often some hangup that eventually makes each debugger ineffective in some circumstance (so I switch to a different one for that task). Every interactive debugger has failed to be effective at some point, so I end up falling back to logging (printf). Logging has been the most robust, even when faced with timing issues in highly concurrent systems (which doesn't lend itself to interactive debugging).

There are ways to debug multi-threaded python programs, but I thought it would simplest if I supported pdb with a single thread mode so that ya may keep using pdb. If you enjoy pdb, keep using it :).

My tip on debugging is: expect that you'll need to know multiple methods of (or tools for) debugging.
